packages/core/src/pubsub/__tests__/pubsub-keepalive.test.ts

```diff
@@ -12,35 +13,106 @@ const diagnosticsLogger = loggerProvider.getDiagnosticsLogger()
 
 const PEER_ID = 'PEER_ID'
 
-test('publish keepalive', async () => {
-  const ipfs = {
-    pubsub: {
-      subscribe: jest.fn(),
-      unsubscribe: jest.fn(),
-      ls: jest.fn(() => []),
-      publish: jest.fn(),
-    },
-    id: jest.fn(async () => ({ id: PEER_ID })),
-  }
-  const keepaliveInterval = 100
-  const pubsub = new Pubsub(
-    ipfs as unknown as IpfsApi,
-    TOPIC,
-    3000,
-    pubsubLogger,
-    diagnosticsLogger
-  )
-  const pubsubWithKeepalive = new PubsubKeepalive(pubsub, keepaliveInterval)
-  const subscription = pubsubWithKeepalive.subscribe()
+describe('pubsub keepalive', () => {
```

Comment:
one case that's still missing is making sure that a user publishing a random non-keepalive message prevents the need to send keepalives in the maxPubsubIntervalTime, but we'll still need to send one if there's no keepalive messages specifically within the maxKeepaliveIntervalTime

Reply:
that said this PR has gone on long enough, I think the coverage you added thus far is good enough for now.  So consider that an optional suggestion or future improvement.